<a href="https://colab.research.google.com/github/Rohitv97/Data-Science-CE888/blob/master/Lab5/my_recommender.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# **Jester Dataset**
1. This is only part of the entire dataset. This contains 24983 rows and 101 columns.
2. Each row corresponds to a user.
3. First column is number of jokes rated by that user
4. Others are ratings for each joke. Ratings are in the range of -10 to 10. Value of 99 indicates Null. That is no rating.

All information was gained by looking at the dataset and from the webpage of the [dataset](http://eigentaste.berkeley.edu/dataset/). 


In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/albanda/CE888/master/lab5-recommender/jester-data-1.csv', header=None)
df

0      1      2      3      4    ...    96     97     98     99     100
0       74  -7.82   8.79  -9.66  -8.16  ...  99.00  -5.63  99.00  99.00  99.00
1      100   4.08  -0.29   6.36   4.37  ...  -2.14   3.06   0.34  -4.32   1.07
2       49  99.00  99.00  99.00  99.00  ...  99.00  99.00  99.00  99.00  99.00
3       48  99.00   8.35  99.00  99.00  ...  99.00  99.00  99.00  99.00  99.00
4       91   8.50   4.61  -4.17  -5.39  ...   1.55   3.11   6.55   1.80   1.60
...    ...    ...    ...    ...    ...  ...    ...    ...    ...    ...    ...
24978  100   0.44   7.43   9.08   2.33  ...   9.03   6.55   8.69   8.79   7.43
24979   91   9.13  -8.16   8.59   9.08  ...  -8.20  -7.23  -8.59   9.13   8.45
24980   39  99.00  99.00  99.00  99.00  ...  99.00  99.00  99.00  99.00  99.00
24981   37  99.00  99.00  99.00  99.00  ...  99.00  99.00  99.00  99.00  99.00
24982   72   2.43   2.67  -3.98   4.27  ...  99.00  99.00  99.00  99.00  99.00

[24983 rows x 101 columns]

In [3]:
## Copy values and keep
test_data = df.copy()
df = df.loc[:, 1:]  ##first column is just number of jokes, and not ratings
arr = df.values
print(arr.shape)

(24983, 100)


In [4]:
## Take all rated jokes. If not rated don't add.
## np.where returns an ndarray with all elements which satisfy the condition (x), and then all elements which don't satisfy the condition
## Over here first array is all rows, and 2nd array is all columns indices where it is not 99
## Use this to understand, why it is used inside the Replace Function
rated = np.where(arr!=99)
print(len(rated[0]), rated[1].shape)

1810455 (1810455,)


In [0]:
def replace(orig, percentage=0.1):
  """
  Replaces 'percentage'% of the original values in 'orig' with 99's
  :param orig: original data array
  :param percentage: percentage of values to replace (0<percentage<1)
  """
  new_data = orig.copy()
  rated = np.where(arr!=99)
  n_rated = len(rated[0])
  idx = np.random.choice(n_rated, size=int(percentage*n_rated), replace=False)
  new_data[rated[0][idx], rated[1][idx]] = 99
  return new_data, (rated[0][idx], rated[1][idx])

In [0]:
new_arr, idx = replace(arr, 0.1)

In [88]:
print(idx)
print(idx[0][1], idx[1][1])

(array([24320, 24804, 20872, ..., 20468, 21321, 24034]), array([69, 58, 89, ..., 30, 20, 25]))
24804 58


# **Latent Factor**

# Notes on training

1. It takes approx 36 seconds to train 1 iteration. 
300,000 iterations is possible only if enough computation power is available.

2. As of now, for normal training, after 9th iteration, the MSE starts increasing again. This could be due to overfitting. Try modifying hyperparameters to find convergence and stopping criteria.

# Training results

Alpha = 0.0001 -> around 9 iterations. (lowest MSE = 426.) Increases after this.

Alpha = 0.001 & Decay = 0.1 -> around 7 iterations (lowest MSE = 402.) Starts increasing after this value.

Alpha = 0.001 & Regularization = 0.1  and Decay = 0.1-> around 7 iterations (lowest MSE = 404). Increases after this.

Alpha = 0.001 & Regularization = 0.1  and Decay = 0.1. Also using more latent features (10). It was 4 before this. -> around 12 iterations. (lowest MSE = 366). Increases after this.

Alpha = 0.001 & Regularization = 0.1  and Decay = 0.1. 20 LF -> around 15 iterations. (lowest MSE = 299). Increases after this.

**This means we can use more Latent Features. Theoretically, more LFs should result in a lower MSE, with the trade off being computation power and time.**

In [0]:
n_latent_factors = 50

user_ratings = new_arr
# Initialise as random values
latent_user_preferences = np.random.random((user_ratings.shape[0], n_latent_factors))
latent_item_features = np.random.random((user_ratings.shape[1], n_latent_factors))

In [0]:
def predict_rating(user_id, item_id):
    """ Predict a rating given a user_id and an item_id.
    """
    user_preference = latent_user_preferences[user_id]
    item_preference = latent_item_features[item_id]
    return user_preference.dot(item_preference)


def train(user_id, item_id, rating, alpha=0.0001, weight_decay=0.1):
    
    #print item_id
    prediction_rating = predict_rating(user_id, item_id)
    err =  prediction_rating - rating
    #print err
    user_pref_values = latent_user_preferences[user_id][:]
    latent_user_preferences[user_id] -= alpha * err * (latent_item_features[item_id])
    latent_item_features[item_id] -= alpha * err * (user_pref_values)
    return err
    

def sgd(iterations=100):
    """ Iterate over all users and all items and train for 
        a certain number of iterations
    """
    for iteration in range(iterations):
        error = []
        for user_id in range(latent_user_preferences.shape[0]):
            for item_id in range(latent_item_features.shape[0]):
                rating = user_ratings[user_id][item_id]
                if (rating!=99):
                # if not np.isnan(rating):
                    err = train(user_id, item_id, rating)
                    error.append(err)
        mse = (np.array(error) ** 2).mean()   
        # if (iteration % 100) == 0:
        #     print('MSE ', iteration, ': ', mse)
        print('MSE ', iteration, ': ', mse)


In [38]:
sgd()

MSE  0 :  28.371271027917963
MSE  1 :  25.015122496766043
MSE  2 :  24.928617926547922
MSE  3 :  24.859419332998346
MSE  4 :  24.799191005829137
MSE  5 :  24.740883428413046
MSE  6 :  24.677474671788037
MSE  7 :  24.600821765960504
MSE  8 :  24.500482956117423
MSE  9 :  24.36252042159121
MSE  10 :  24.168562271179628
MSE  11 :  23.895898066889128
MSE  12 :  23.520090330319395
MSE  13 :  23.021998504031398
MSE  14 :  22.39974911161156
MSE  15 :  21.681305682977587
MSE  16 :  20.926911020168166
MSE  17 :  20.212212621699134
MSE  18 :  19.59878373609533
MSE  19 :  19.113551662074634
MSE  20 :  18.750327093312162
MSE  21 :  18.48542212803414
MSE  22 :  18.292347350029992
MSE  23 :  18.149074649333958
MSE  24 :  18.03950309751691
MSE  25 :  17.952486019377837
MSE  26 :  17.880421341137655
MSE  27 :  17.81809748266523
MSE  28 :  17.761888199467926
MSE  29 :  17.70922128585818
MSE  30 :  17.658233519625828
MSE  31 :  17.607546922238065
MSE  32 :  17.556123563293
MSE  33 :  17.50317159041398
M

In [0]:
def validation_check():
  error = []
  prediction = latent_user_preferences.dot(latent_item_features.T)
  for i in range(len(idx[0])):
    predicted_rating = prediction[idx[0][i]][idx[1][i]]
    actual_rating = arr[idx[0][i]][idx[1][i]]
    err = predicted_rating - actual_rating
    error.append(err)
  mse = (np.array(error) ** 2).mean()
  print('Validation MSE: ', mse)
  

In [94]:
validation_check()

Validation MSE:  13.206327761458075


In [96]:
### Creating new validation arrays from original data randomly again
for i in range(10):
  new_arr, idx = replace(arr, 0.1)
  validation_check()

Validation MSE:  13.23793563989452
Validation MSE:  13.325165122837326
Validation MSE:  13.207294623008817
Validation MSE:  13.180494257278031
Validation MSE:  13.179841998969904
Validation MSE:  13.244275441806645
Validation MSE:  13.129248377382421
Validation MSE:  13.313002542228993
Validation MSE:  13.26293492570781
Validation MSE:  13.271015821681216


# Results

## LF: 100, Iterations: 100
Each trial makes use of a different validation dataset

Training Error = 12.86
1. Validation Error = 13.206
2. Validation Error = 13.237
3. Validation Error = 13.207
4. Validation Error = 13.180
5. Validation Error = 13.179
6. Validation Error = 13.244
7. Validation Error = 13.129
8. Validation Error = 13.313
9. Validation Error = 13.262
10. Validation Error = 13.271
11. Validation Error = 13.325

Average Validation Error = 13.232







In [0]:
prediction = latent_user_preferences.dot(latent_item_features.T)

In [101]:
### Prediction values on entire Test Dataset

values = [zip(df.values[i], prediction[i]) for i in range(prediction.shape[0])]
comparison_data = pd.DataFrame(values)
comparison_data.columns = df.columns
comparison_data

1    ...                           100
0      (-7.82, -1.1972836166847933)  ...    (99.0, -5.675803996306302)
1          (4.08, 3.63277210446559)  ...     (1.07, 7.649424818976035)
2         (99.0, 9.588316355264014)  ...    (99.0, 6.5638220300881995)
3        (99.0, 3.3328060800885475)  ...     (99.0, 8.267518079974518)
4         (8.5, 1.0737923652773484)  ...     (1.6, 3.6421660286881536)
...                             ...  ...                           ...
24978     (0.44, 4.655072751178377)  ...    (7.43, 10.211828996848393)
24979     (9.13, 4.243598809392026)  ...     (8.45, 5.005666862018283)
24980     (99.0, 1.613125392470121)  ...  (99.0, -0.08095070580414203)
24981    (99.0, 1.4723414566811381)  ...   (99.0, -1.6507162209694384)
24982     (2.43, 2.146157064382971)  ...    (99.0, 0.8671898264035316)

[24983 rows x 100 columns]

In [109]:
jokes_ratings = pd.DataFrame(prediction)
mean_idx = jokes_ratings.mean()
min_idx = mean_idx.idxmin()
max_idx = mean_idx.idxmax()
print('Worst Joke ID predicted: ', min_idx, '; Average Rating: ', mean_idx[min_idx])
print('Best Joke ID predicted: ', max_idx, '; Average Rating: ', mean_idx[max_idx])

Worst Joke ID predicted:  57 ; Average Rating:  -4.004488290910803
Best Joke ID predicted:  49 ; Average Rating:  3.5214865773076363
